### PR TITLE
Store Patient Filters in Redux Reducer

### PR DESCRIPTION
**Pull request by ericsheng495**
## [Frontend-ish] Store Patient Filters in Redux Reducer

Issue Number(s): #90 

What does this PR change and why?
Currently, we store all the patient search (page - `/patient/search`; file - `/src/app/(management-portal)/patient/search/page.tsx`) filters as states in the page. We want a way for these filters to persist across page reloads and across different pages (eg: new feature - group analytics).

### Checklist
- [x] Create `patientSearchReducer` with appropriate actions
- [x] Refactor the code to use this reducer, using `useDispatch` and `useState`

### Critical Changes

- Database change/migration to run
- Environment config change
- Breaking API change

### Related PRs

- #number_of_pr

### Testing

Enumerate steps to test the functionality of your ticket. This should include edge cases and testing of error handling, if applicable.

### CODE REVIEW

File: src/app/(management-portal)/patient/search/page.tsx [hunk 1]
```diff
@@ -90,6 +91,7 @@ export default function Page() {
     secondaryNames,
     sortField,
     currentPage,
+    dispatch,
```
Comment (samarth52) [inline src/app/(management-portal)/patient/search/page.tsx]: `dispatch` is not used in this `useEffect`, so it should not be present in the dependency list.

File: src/app/(management-portal)/patient/search/page.tsx [hunk 2]
```diff
@@ -108,6 +110,7 @@ export default function Page() {
     secondaryPhoneNumbers,
     secondaryNames,
     sortField,
+    dispatch,
```
Comment (samarth52) [inline src/app/(management-portal)/patient/search/page.tsx]: `dispatch` is not used in this `useEffect` as well, so it should not be present in the dependency list.

File: src/components/Search/AdvancedSearch/AdvancedSearch.tsx [hunk 3]
```diff
@@ -91,23 +91,8 @@ export const AdvancedSearch = (props: UpdateParamProp) => {
...
-          const newSet = new Set<T>(set);
-          return newSet.add(element);
-        }
-        return set;
-      });
-    },
-    [],
-  );
+  // eslint-disable-next-line @typescript-eslint/no-shadow
+  const { active } = useSelector((state: RootState) => state.patientSearch);
```
Comment (samarth52) [inline src/components/Search/AdvancedSearch/AdvancedSearch.tsx]: Do not use eslint disable commands unless the errors are unavoidable. Instead of using the eslint disable command here, you can rename the `state` parameter to something else.

File: src/components/Search/AdvancedSearch/AdvancedSearch.tsx [hunk 4]
```diff
@@ -91,23 +91,8 @@ export const AdvancedSearch = (props: UpdateParamProp) => {
...
-          !set.has(element)
-        ) {
-          const newSet = new Set<T>(set);
-          return newSet.add(element);
-        }
-        return set;
-      });
-    },
-    [],
-  );
```
Comment (samarth52) [inline src/components/Search/AdvancedSearch/AdvancedSearch.tsx]: This function should not be deleted and should instead be refactored. Those fields whose states are of a `Set` type can accept multiple filters. For example, `emails` is of type `Set<string>` which means that the user can add a new email filter every single time they click "Apply". I can enter something like "hello@gmail.com" in the email field and press Apply. Then I can subsequently enter "test@gmail.com" in the email field and press Apply. The results shown in the table will then show all records whose `email` field matches either "hello@gmail.com" or "test@gmail.com". There will also be a `Tag` for each of these emails. Please refer to the original functionality and make this change.

File: src/components/Search/AdvancedSearch/AdvancedSearch.tsx [hunk 5]
```diff
@@ -117,21 +102,24 @@ export const AdvancedSearch = (props: UpdateParamProp) => {
...
+    if (city) dispatch(setCities(new Set([city])));
+    if (dateOfBirth) dispatch(setDateOfBirths(new Set([dateOfBirth])));
+    if (email) dispatch(setEmails(new Set([email])));
+    if (additionalAffiliation)
+      dispatch(setAdditionalAffiliations(new Set([additionalAffiliation])));
+    if (dateOfJoin) dispatch(setDateOfJoins(new Set([dateOfJoin])));
+    if (beiChapter) dispatch(setBeiChapters(new Set([beiChapter])));
+    if (secondaryPhoneNumber)
+      dispatch(setSecondaryPhoneNumbers(new Set([secondaryPhoneNumber])));
+    if (secondaryName) dispatch(setSecondaryNames(new Set([secondaryName])));
```
Comment (samarth52) [inline src/components/Search/AdvancedSearch/AdvancedSearch.tsx]: Refer to my previous comment about the `checkAndUpdateList` function. For the states of `Set` type, you should add the new value to the `Set` and not replace the previous value(s).

**Review by samarth52 — CHANGES_REQUESTED** [review]
Great work on the refactor! There are a few changes to be made which I have noted in my review comments. But otherwise, great job!

**Review by samarth52 — APPROVED** [review]
lgtm 👍 Great job!
